Server-side pagination - First, limited change, disabled in code

richard-cox opened the pull request:
<!-- This template is for Devs to give QA details before moving the issue To-Test -->
### Summary
Contributes to rancher/dashboard#9545
<!-- Define findings related to the feature or bug issue. -->


### Occurred changes and/or fixed issues
- Wire in limited integration to support server side pagination in ResourceList component
- Feature disabled in code
  - Pagination can be globally disabled, disabled per store, disabled per resource or if resource is generic
  - At the moment pagination is hardcoded to be OFF
    - **see shell/utils/pagination-utils.ts for hardcoded settings, set top level `enabled: true` to use the feature**
- Once enabled in code, applies to ...
  - Cluster configmap, secret and pod resources
     - Sorting / Filtering is disabled for any column who's value was computed locally (e.g. model getters)
       - secret - subTypeDisplay - translates things like `kubernetes.io/service-account-token` to `Svc Acct Token`
       - configmaps - data - we convert data from two properties into something more readable
     - For these types to work custom steve pagination headers are used that reference the object itself and nothing (ish) from the model
  - Cluster lists that have no custom headers or component
- Paginated lists do not subscribe for updates, user has to press a Refresh button to see changes (or change pag settings page, filter, sort, etc)
  - If a resource outside of the current page changes it is ignored and not saved to the store. Any resource that needs the latest will pick it up automatigically at the time the next findX is made
- Every pagination request fetches the latest results, resource revision is ignored
- Manual refresh should continue to work if enabled, however will be removed once pagination is on by default
- Incremental loading is not compatible with pagination (though that work is in part 2)
- The Namespace filter 'Not in a Project' is not compatible with pagination (it would involve a huge projectsornamespaces param)
- It should be very easy to bring in pagination to non-cluster stores, or non rancher api's 

#### Pending Changes
To avoid this PR growing initial trial pagination will come in parts

_Part 1 Initial changes, hardcoded to be off_

This PR, see description

_Part 2 Integrate Performance Settings_
- Currently hardcoded in shell/utils/pagination-utils.ts. Once settings are confirmed they can be added to `Performance` options
- Pagination dependent on vai feature flag being on
- Pagination incompatible with incremental loading and enforced project/namespace filtering settings
- Incremental Loading is incompatible with vai backed api. requires feature flag being off (limit no longer supported)
- Discussion - Enabled/Disable given resource count?

_Part 3 Loading Indicators, and Updating results in a list_
- Update loading indicator following UX discussion. This is currently horrible and jarring
- Confirm manual refresh / reload button is go forward. If so we don't need to handle socket messages for the type and code can be removed

_Part 3 Changes following API Updates_
- There are backend fixes (detailed below) that need validating / changes applying. We can only do that though when we get them from backend team
- There may be other changes as we work through this with them, specifically handling non-paginated requests their revision numbers. We will need to react to them as and when they happen

_Part 4 Automated Tests_
- Once everything has settled a selection of e2e and unit tests will be written


### Technical notes summary

Notes on **existing** list architecture 
- Component hierarchy (always one of)
  - ResourceList --> ResourceTable --> SortableTable
    - ResourceList has the resource-fetch mixin which handles fetch of resources, manual refresh, etc
  - ResourceList --> CustomList --> ResourceTable --> SortableTable
    - ResourceList has the resource-fetch mixin
    - CustomList normally ALSO has the resource-fetch mixin

General Configuration
- Overall settings that enable/disable pagination can be found in shell/utils/pagination-utils.ts (these will become performance settings)
- Generic lists that have no header config that use schema's to create them... are supported and no other changes required
- Resources that have custom headers or a custom list component have to be converted one by one. Configmaps, secrets and pods have been converted
- Converted resources provide pagination headers in the same way they would provide normal headers
- Pagination headers should reference specific paths within the resource rather than computed model properties

General flow
- Previously pagination was handled withing SortableTable. Not it's all about pushing in a page to SortableTable and it pushing out pagination settings
- So the kinda flow is..
  - a new mixin used in resource list mixin determines if a list supports pagination, bubbles down pagination settings to sortable table
  - sortable table bubbles up pagination changes 
  - resource list pagination mixin takes the changes and makes request for new page 
    - find action takes pagination settings
      - pagination settings applied in urlFor
    - find action persists page, pagination settings and pagination results (total results)
  - resource list bubbles down pagination results (total results) to sortable table


General Improvements
- Mutations: when `loadX` functions call `load` a lot of times it was recalculating some common args. These are now done one up front

    
#### Backend / API Issues 

> TODO: RC These need validating with new image

 - configmaps
   - different `count` provided when adding/removing `sort=-metadata.namespace`
     - configmaps?page=1&pagesize=10&sort=-metadata.namespace,-metadata.name,-id&filter=metadata.name=kube. count:14
     - configmaps?page=2&pagesize=10&sort=-metadata.namespace,-metadata.name,-id&filter=metadata.name=kube. count:14
     - configmaps?page=1&pagesize=10&sort=-metadata.name,-id&filter=metadata.name=kube,metadata.namespace=kube. count:18
     - configmaps?page=2&pagesize=10&sort=-metadata.name,-id&filter=metadata.name=kube,metadata.namespace=kube. count:18
- pod
  - some columns come from schemas. the field is a specific value in an array. can we sort by it?
    - pod 'restarts' field is determined from a reference in it's schema. `attributes.columns[3]` `field` points to roughly the path in the pod object where the value is `$.metadata.fields[3]"`.
    - we can filter by this (`filter=metadata.fields.3=xyz`)
    - sorting though doesn't seem to work (`sort: metadata.fields.3`)
- secrets
  - `_type` now comes from `metadata.fields.1`. faces same problem as pods above     
     
### Areas or cases that should be tested
- ConfigMaps, Secrets and Pod list work correctly
  - User can page, filter, sort, etc
 - user navigates to last page in a list with a single result on the last page --> in another browser delete the resource
- user navigates to page before last in a list with a single result on the last page --> in another browser delete the resource --> nav to last page
- no regressions for
  - non-paginated lists
  - namespace/project filtering
  - manual refresh (list)
  - incremental loading (list)
  - selection of workspace in fleet
  - List group by nothing / namespace / (pod node)
    - group by namespace sticks on refresh
    - The setting to hide system resource (aka local cluster c-/p- namespaces)
  - Norman lists, Cluster Management
  - Updates correctly change content in pag, sort, etc

richard-cox commented on shell/components/ResourceList/index.vue:
```diff
@@ -225,11 +255,14 @@ export default {
       :adv-filter-prevent-filtering-labels="advFilterPreventFilteringLabels"
       :use-query-params-for-simple-filtering="useQueryParamsForSimpleFiltering"
       :force-update-live-and-delayed="forceUpdateLiveAndDelayed"
+      :external-pagination="canPaginate"
```
external-pagination-enabled

richard-cox commented on shell/plugins/steve/steve-pagination-utils.ts:
```diff
@@ -0,0 +1,84 @@
+      filter: bFilter, sort: bSort = [], namespaces: bNamespaces = [], ...bPrimitiveTypes
+    } = b || {};
+
+    if (isEqual(aPrimitiveTypes, bPrimitiveTypes) &&
```
just return

richard-cox commented:
@codyrancher I've gone through a couple of rounds with Neil and he suggested you might like a look as well. There's a lot in here to unpack, happy to give an overview via a call. Just let me know

codyrancher commented on shell/components/ResourceList/index.vue:
```diff
@@ -124,7 +128,7 @@ export default {
       }
 
-      return this.$store.getters['type-map/headersFor'](this.schema);
+      return this.$store.getters['type-map/headersFor'](this.schema, this.canPaginate);
```
By any chance will pagination be a part of the type schema? If so this probably isn't necessary.

richard-cox replied:
The BE will support pagination of all steve types, though this kind of stuff is because we live in a halfway world. Ultimately almost all resources will be paginated. If we ever get to the stage where all are we can rip this kind of thing out.

codyrancher commented on shell/components/ResourceList/index.vue:
```diff
@@ -148,14 +153,29 @@ export default {
     namespaceFilter(neu, old) {
-      if (sameContents(neu, old)) {
-        return;
+      if (neu && !this.componentWillFetch) {
```
Previously it looks like if the user set namespaceFilter to something falsy like null (would this be null for All Namespaces?) it would trigger a fetch. Are we intentionally ignoring this case?

richard-cox replied:
This property relates to the `Require Namespace / Project Filtering` feature. In that world we won't show the list if the user hasn't selected any namespaces or projects, so won't need to kick off a request. So this should be a safe change. Just in case i gave it a test going from config map list with `All Namespaces` selected --> enable feature --> config list and and no list was shown --> select a project --> list shown with correct rows

codyrancher commented on shell/components/ResourceTable.vue:
```diff
@@ -184,24 +194,31 @@ export default {
 
   computed: {
+    options() {
+      return this.$store.getters[`type-map/optionsFor`](this.schema, !!this.externalPaginationEnabled);
```
Why !!this.externalPaginationEnabled, it looks like externalPaginationEnabled is already  boolean so a conversion shouldn't be needed.

richard-cox replied:
Good catch. At one point it wasn't, will update

codyrancher commented on shell/components/ResourceTable.vue:
```diff
@@ -344,7 +365,12 @@ export default {
+          // and if not that just fall back to the default
+          const defaultGroup = this.groupOptions.find((g) => g.value === DEFAULT_GROUP);
+
+          return defaultGroup ? DEFAULT_GROUP : this.groupOptions[0]?.value || DEFAULT_GROUP;
```
Using operator short circuiting in the same line of a ternary is a little difficult to parse and  determine the intent.

I imagine another an additional assignment could obviate the need for the above comment and make the return line a little less ambiguous.

richard-cox replied:
Good point. Have updated to the following 

```
          // Attempt to find the default option in available options...
          // if not use the first value in the options collection...
          // and if not that just fall back to the default
          if (this.groupOptions.find((g) => g.value === DEFAULT_GROUP)) {
            return DEFAULT_GROUP;
          }

          return this.groupOptions[0]?.value || DEFAULT_GROUP;
```

codyrancher commented on shell/components/SortableTable/index.vue:
```diff
@@ -282,6 +284,14 @@ export default {
+    /**
+     * The list will always be sorted by these regardless of what the user has selected
+     */
+    mandatorySort: {
```
I didn't see an obvious use case for this. Shouldn't we disable sorting in scenarios where this would be used?

codyrancher commented on shell/config/product/explorer.js:
```diff
@@ -198,20 +204,47 @@ export function init(store) {
+    ]
+  );
+
+  const secretData = {
+    name:      'data',
+    labelKey:  'tableHeaders.data',
+    value:     'dataPreview',
+    formatter: 'SecretData'
+  };
```
Do we want this (along with the other types defined in this file) to be with table-headers?

codyrancher commented on shell/mixins/resource-fetch-api-pagination.js:
```diff
@@ -0,0 +1,237 @@
+import {
```
I don't know if we want to continue adding new mixins

![image](https://github.com/rancher/dashboard/assets/55104481/6fa69e33-db87-4a68-aa32-59edaa753f0a)

richard-cox replied:
I was 50:50 on this. Went with a mixin as there's an existing set and pattern for ResourceList. These are all tightly coupled and depend on each other.
- `resource-fetch`
  - contains the route $fetchType which makes api requests used in ResourceList and custom list components. This references the other mixins and is used by them to make that request.
- `resource-fetch-namespaced`
  - supplies functionality for `Require Namespace / Project Filtering` to sibling mixins and parent components
- `resource-fetch-api-pagination` (new one)
  - supplies functionality for pagination to sibling mixins and parent components
  
When we get to the point no incremental loading, manual refresh or required ns / filtering is needed we can get rid of the existing mixins and port only the pagination feature over

codyrancher commented on shell/plugins/steve/getters.js:
```diff
@@ -35,48 +41,79 @@ export default {
-          return `${ encodeURI(key) }=${ encodeURI(val) }`;
+          // Steve's filter options now support more complex filtering not yet implemented here #9341
+          if (isSteve) {
+            url += `${ (url.includes('filter=') ? '&' : 'filter=') }`;
```
It looks like alot of these changes are really just nesting so you probably don't want to touch them but in case you did the url manipulation in this file looks like it should either use stuff or move stuff to the utils/url module.

richard-cox replied:
yeah i cringed at this fn. the core change was just moving things around. i had thought about converting everything over to something like `<URL>.searchParams.append` but thought the PR was a big enough change as is.

codyrancher reviewed (dismissed):
Even though I have comments here I think it would be better to address them in future PRs if you choose to. I see that most of the functionality is working reasonably well if you enable pagination so feel free to push.


Here's a few things I noticed while testing:

Do we want to preserve the page the user was on while navigating back?

https://github.com/rancher/dashboard/assets/55104481/97fff179-8508-4489-8e61-7a6d0e837f90

Deleting items doesn't update the page numbers or the ability to navigate to pages that no longer exist

![Screenshot 2024-03-26 205824](https://github.com/rancher/dashboard/assets/55104481/015feff7-0c1e-45be-8c0a-081d34d295b9)
![Screenshot 2024-03-26 205843](https://github.com/rancher/dashboard/assets/55104481/227b5615-4093-4b55-bf41-dcd1d6d4399f)
![Screenshot 2024-03-26 205930](https://github.com/rancher/dashboard/assets/55104481/60bfa22b-3a6d-4ea7-ab5f-2f0cb31e9fcc)

https://github.com/rancher/dashboard/assets/55104481/8bf0a9d9-5b23-48ac-9ec0-5d50062985a4

The navigation count didn't match the load/pagination totals (I see this in master too so I don't think you caused the issue here):

![Screenshot 2024-03-26 205723](https://github.com/rancher/dashboard/assets/55104481/60e939ec-1a8c-4809-9aa4-7897aec5c04f)

codyrancher reviewed (approved):
As long as tests pass LGTM.

richard-cox commented:
@codyrancher Should be ready to review again. I've moved a lot around to fix / improve things. See commit comment for details. Most of it revolves around the local cluster where namespaces for projects exist... as well as the project itself. They have the same id and using it in new `projectsornamespaces` param would return a combination of user (project) and system (project's namespace) resources... which was breaking the user / system filter. Also spotted the 'show all namespaces' user preference wasn't correctly applied and fixed that too

richard-cox commented on shell/types/store/dashboard-store.types.ts:
```diff
@@ -1,29 +1,161 @@
+ *
+ * For more information see https://github.com/rancher/steve?tab=readme-ov-file#filter
+ */
+export class OptPaginationFilter {// implements OptPaginationFilter
```
@codyrancher I don't think i've got the definitions for this and OptPaginationFilterField right. I'm making use of them a lot more in another context (converting the node's list and it's secondary resources over to pagination world) and i'd like to make them easier to use... but will do it in that PR. If there are no objections can we save the review of these until that one?
